### PR TITLE
Documentation for net_get_interfaces()

### PR DESCRIPTION
**Pull request by tored**
Relates to #663 
Resolves https://bugs.php.net/bug.php?id=79251

Unsure where to put documentation for this function because it is a brand new file (`ext/standard/net.c`) in `php-src`, but I put it into **Other Services / Network / Network Functions**

Requires  more documentation regarding flags and family attributes for the unicast subarray.

### CODE REVIEW

File: reference/network/functions/net-get-interfaces.xml [hunk 1]
```diff
@@ -0,0 +1,144 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- $Revision$ -->
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:2]: This comment can be removed.

File: reference/network/functions/net-get-interfaces.xml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+ <refnamediv>
+  <refname>net_get_interfaces</refname>
+  <refpurpose>Get network interfaces</refpurpose>
+ </refnamediv>
+
+ <refsect1 role="description"><!-- {{{ -->
+  &reftitle.description;
+  <methodsynopsis role="procedural">
+   <type class="union"><type>array</type><type>false</type></type><methodname>net_get_interfaces</methodname>
+   <void />
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:13]:
```suggestion
   <void/>
```

File: reference/network/functions/net-get-interfaces.xml [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+     </tbody>
+    </tgroup>
+   </table>
+  </para>
+ </refsect1><!-- }}} -->
+
+ <refsect1 role="errors"><!-- {{{ -->
+  &reftitle.errors;
+  <para>
+   Issues an E_WARNING on failure to get interface information.
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:119]:
```suggestion
   Emits an <constant>E_WARNING</constant> on failure to get interface information.
```

File: reference/network/functions/net-get-interfaces.xml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+   </table>
+  </para>
+  <para>
+   <table>
+    <title>Unicast attributes</title>
+    <tgroup cols="2">
+     <thead>
+      <row>
+       <entry>Attribute</entry>
+       <entry>Meaning</entry>
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:82]:
```suggestion
       <entry>&Name;</entry>
       <entry>&Description;</entry>
```

File: reference/network/functions/net-get-interfaces.xml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+   This function returns an associative array where key  is the name of the interface and
+   the value an associative array of interface attributes,
+   &return.falseforfailure;. Each interface associative array contains
+   <table>
+    <title>Interface attributes</title>
+    <tgroup cols="2">
+     <thead>
+      <row>
+       <entry>Attribute</entry>
+       <entry>Meaning</entry>
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:37]:
```suggestion
       <entry>&Name;</entry>
       <entry>&Description;</entry>
```

File: reference/network/functions/net-get-interfaces.xml [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+ <refsect1 role="parameters"><!-- {{{ -->
+  &reftitle.parameters;
+  &no.function.parameters;
+ </refsect1><!-- }}} -->
+
+ <refsect1 role="returnvalues"><!-- {{{ -->
+  &reftitle.returnvalues;
+  <para>
+   This function returns an associative array where key  is the name of the interface and
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:28]:
```suggestion
   Returns an associative &array; where the key is the name of the interface and
```

File: reference/network/functions/net-get-interfaces.xml [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+   the value an associative array of interface attributes,
+   &return.falseforfailure;. Each interface associative array contains
```
Comment (Girgias) [inline reference/network/functions/net-get-interfaces.xml:30]:
```suggestion
   &return.falseforfailure;.
  </para>
  <para>
   Each interface associative array contains:
```

**Review by Girgias — CHANGES_REQUESTED** [review]
I don't really now how the function is meant to behave but a couple of nits.

**Comment by tored** [comment]
> I don't really now how the function is meant to behave but a couple of nits.

My pull request isn't 100% done, but the function is a bit weird, the outer structure is for POSIX, and the inner structure (unicast) is mostly Windows. Mixing both results in the same documentation can be confusing, because some keys  is not even present on the other platform. Name for the adapter differs too, if I remember correctly it is a GUID on Windows but a string name on POSIX. And I think some field refers to a platform specific constant.

Should we separate the result for POSIX and Windows? Or should we still try to keep everything together but add notes that some properties does not exist on the other platform? Can be messy.  I personally think that separating POSIX and Windows two different tables is better to avoid the confusion, but I'm not aware if php-doc does this for other functions.

How do you usually do this in the doc group?



The outer structure is mainly for

**Comment by tored** [comment]
Okay, I remembered wrong, it is mostly Windows that uses the outer structure.

Here is an example for Linux

```
array(2) {
  ["lo"]=>
  array(2) {
    ["unicast"]=>
    array(3) {
      [0]=>
      array(2) {
        ["flags"]=>
        int(65609)
        ["family"]=>
        int(17)
      }
      [1]=>
      array(4) {
        ["flags"]=>
        int(65609)
        ["family"]=>
        int(2)
        ["address"]=>
        string(9) "127.0.0.1"
        ["netmask"]=>
        string(9) "255.0.0.0"
      }
      [2]=>
      array(4) {
        ["flags"]=>
        int(65609)
        ["family"]=>
        int(10)
        ["address"]=>
        string(3) "::1"
        ["netmask"]=>
        string(39) "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"
      }
    }
    ["up"]=>
    bool(true)
  }
  ["enp0s3"]=>
  array(2) {
    ["unicast"]=>
    array(3) {
      [0]=>
      array(2) {
        ["flags"]=>
        int(69699)
        ["family"]=>
        int(17)
      }
      [1]=>
      array(5) {
        ["flags"]=>
        int(69699)
        ["family"]=>
        int(2)
        ["address"]=>
        string(9) "10.0.2.15"
        ["netmask"]=>
        string(13) "255.255.255.0"
        ["broadcast"]=>
        string(10) "10.0.2.255"
      }
      [2]=>
      array(4) {
        ["flags"]=>
        int(69699)
        ["family"]=>
        int(10)
        ["address"]=>
        string(24) "fe80::a00:27ff:fe35:9693"
        ["netmask"]=>
        string(21) "ffff:ffff:ffff:ffff::"
      }
    }
    ["up"]=>
    bool(true)
  }
}
```

Windows

```
array(2) {
  ["{AE348A58-926B-4BE4-ADA1-0E4F74BF1A4A}"]=>
  array(5) {
    ["description"]=>
    string(33) "Microsoft Hyper-V Network Adapter"
    ["mac"]=>
    string(17) "00:15:5D:0B:27:88"
    ["mtu"]=>
    int(1500)
    ["unicast"]=>
    array(2) {
      [0]=>
      array(4) {
        ["flags"]=>
        int(453)
        ["family"]=>
        int(23)
        ["address"]=>
        string(25) "fe80::dc54:f386:1c72:5604"
        ["netmask"]=>
        string(21) "ffff:ffff:ffff:ffff::"
      }
      [1]=>
      array(4) {
        ["flags"]=>
        int(453)
        ["family"]=>
        int(2)
        ["address"]=>
        string(14) "172.23.193.175"
        ["netmask"]=>
        string(13) "255.255.240.0"
      }
    }
    ["up"]=>
    bool(true)
  }
  ["{D62CD1F7-07AF-11EC-B28D-806E6F6E6963}"]=>
  array(4) {
    ["description"]=>
    string(29) "Software Loopback Interface 1"
    ["mtu"]=>
    int(4294967295)
    ["unicast"]=>
    array(2) {
      [0]=>
      array(4) {
        ["flags"]=>
        int(385)
        ["family"]=>
        int(23)
        ["address"]=>
        string(3) "::1"
        ["netmask"]=>
        string(39) "ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"
      }
      [1]=>
      array(4) {
        ["flags"]=>
        int(385)
        ["family"]=>
        int(2)
        ["address"]=>
        string(9) "127.0.0.1"
        ["netmask"]=>
        string(9) "255.0.0.0"
      }
    }
    ["up"]=>
    bool(true)
  }
}
```

**Comment by cmb69** [comment]
If the results are grossly different on Windows and other platforms, it makes sense to separate the results. Not sure if that's done elsewhere, but function results may not be that different for other functions.

**Comment by cmb69** [comment]
While I think this doc page can be improved, given that the function is available for more than three years, this PR is a reasonable start. Thank you very much! I'm going to tweak a bit (most notably regarding the Windows/POSIX differences, which don't appear to be that grave).

File: reference/network/functions/net-get-interfaces.xml (hunk 1 above)
Comment (cmb69) [reply reference/network/functions/net-get-interfaces.xml:2]: Will do.

**Pull request closed without merging**